### ISSUE TITLE
reportImportCycles is still triggered when something is imported with TYPE_CHECKING

### ISSUE THREAD

**Issue by Wizzerinus** [DetachHead/basedpyright#960]
So here is the test case:
```
% pyright b.py a.py
0 errors, 0 warnings, 0 informations 

~/tests
% basedpyright b.py a.py
/home/wizzerinus/tests/b.py
  /home/wizzerinus/tests/b.py:1:8 - warning: Import "a" is not accessed (reportUnusedImport)
/home/wizzerinus/tests/a.py
  /home/wizzerinus/tests/a.py: error: Cycle detected in import chain
    /home/wizzerinus/tests/a.py
    /home/wizzerinus/tests/b.py (reportImportCycles)
  /home/wizzerinus/tests/a.py:4:12 - warning: Import "b" is not accessed (reportUnusedImport)
1 error, 2 warnings, 0 notes

~/tests
1 % cat b.py
import a

~/tests
% cat a.py 
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import b
```

Using TYPE_CHECKING to stop import cycles is a relatively common idiom so I'd want to see this supported :) I would attempt contributing but it seems that the import cycles rule is pretty difficult to maintain so not sure if starting with that is the best idea ever.

**Comment by DetachHead** [comment]
you can disable the `reportImportCycles` rule or suppress it with a `pyright:ignore` comment if you disagree with its behavior, but this is one of the few cases where i agree with [eric traut's approach](https://github.com/microsoft/pyright/issues/9510#issuecomment-2504972178), so i don't think it's a good idea to reduce its strictness, though i'll try to explain what he means by "architectural layering violations" as i understand it.

the purpose of the `reportImportCycles` rule isn't just to prevent runtime errors caused by circular imports, but to also detect a deeper issue within your your project. when you have a circular dependency, regardless of whether it's a type-only import or not, it very often means there's something wrong with how your project is structured. it means your [concerns are not separated](https://en.wikipedia.org/wiki/Separation_of_concerns).

i know it can be quite difficult to effectively explain and understand *exactly* why something like this is bad when there're no runtime crash, because i've spent months trying to convince coworkers that the circular dependencies in our codebase, even the `TYPE_CHECKING` ones, are 100% avoidable and that eliminating them leads to a much more well-structured and maintainable codebase. so i'll give an example from my experience:

at work we had a single module containing a class with hundreds of global variables that can be passed via the CLI. these variables vary in how specific they are. some of them are generic as in they are relevant to pretty much everything in the project. but some of them are highly specific to a single module, meaning this super generic module that gets imported by everything needs to now import types from a highly specific module that is otherwise rarely imported by anything else:

```py
# utilities/variables.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from specific_module_1 import ModuleSpecificData
    from specific_module_2 import OtherModuleTestData

class Variables:
    def __init__(self):
        ... # logic to parse the variables from the command line

    some_generic_thing: bool
    data_for_some_specific_module: ModuleSpecificData
    data_for_some_other_module: OtherModuleTestData
```
```py
# specific_things/specific_module_1.py
from utilities.variables import Variables

class ModuleSpecificData:
    ...

def create_something():
    result = ModuleSpecificData()
    Variables.data_for_some_specific_module = result
```

in this example, while the circular dependency does not exist at runtime because the `variables.py` module does not actually depend on anything at runtime, there is still a glaring problem with the structure of this project, which is that there's a single, generic module that needs to know about every other module in the project. over time, that module will (and has) grown in size to store hundreds of variables for completely unrelated things that this module should really have no knowledge of. if we ever wanted to split `variables.py` into a separate project (because it contains useful generic functionality for reading variables from the command line), it would be impossible because it's tightly coupled to everything else.

not to mention that the only reason it doesn't crash at runtime is due to the fact that the types *currently* aren't required by that module at runtime, but what if that changes? what if we want to update the class to use [pydantic](https://github.com/pydantic/pydantic) to parse and validate the types of each variable? then we can't use the `TYPE_CHECKING` block anymore so we're screwed.

the solution here is to split it into separate classes, so that only the logic specific to the `variables.py` module is defined there, and the variables used by individual specific modules are defined within those modules:

```py
# utilities/variables.py
class AbstractVariables:
    """extend this class to define variables"""
    def __init__(self):
        ... # logic to parse the variables from the command line
```


```py
# specific_things/specific_module_1.py
from utilities.variables import AbstractVariables

class ModuleSpecificData:
    ...

class SpecificVariables(AbstractVariables):
    data_for_some_specific_module: ModuleSpecificData

def create_something():
    result = ModuleSpecificData()
    SpecificVariables.data_for_some_specific_module = result
```

this, from my understanding, what is meant by "separation of concerns". there are many other beneficial reasons for splitting up concerns like this, some of which vary depending on the language you're using, but in general i consider it to be very important especially when working in a large codebase.

the way i see it, your project is not a collection of modules if it's riddled with circular dependencies, but rather a single module split up arbitrarily into separate files.

**Comment by Wizzerinus** [comment]
The project I am working on is structured in a pretty peculiar way and I don't think it's circular dependencies, although maybe a lot of projects that need the TYPE_CHECKING crutch are.

So basically there are certain objects with bidirectional access to each other, and I want to typecheck them (this isn't the code of the project but it should probably explain what I'm talking about):
```python
# RootObject.py
class RootObject:
  modules: list[Module]
  # 500 lines of code here

# ModuleObject.py
class Module:  # this handles some specific functionality, not a generic class
  root: RootObject
  # 500 lines of code here
```

This pattern appears many times in many unrelated places. Even though this is in some ways a dependency, in some places it is not a dependency at runtime, but one has to be introduced when typechecking is needed:

```python
# GenericObject.py
class GenericObject:
  children: list["GenericObject"]
  parent: "GenericObject"
  def addChild(self, child: "GenericObject"): ...

# ComplexState.py
class ComplexState(GenericObject):
  def __init__(self):
    # this is being composed out of several different sub-objects
    # yeah I get it you can make some `createComplexState()` function to add the submodules
    # but at that point we're breaking the principle of each class being individually useful
    self.addChild(SubObject())

# SubObject.py
class SubObject(GenericObject):
  parent: ComplexState   # oh no!
```
This is a standard composition setup and it does not really require the child nodes knowing about the parent's type, but not having that knowledge just makes typing pretty much impossible. Even if you don't explicitly store the reference, by the way:
```python
class SubObject(GenericObject):
  def executeAction(self, state: "ComplexState"):   # oh no!
```

How I see it the *separation of concerns* should be different from *separation of code files*. I agree about the separation of concerns argument but the examples why this kind of behavior is bad usually consider small code bases, and I believe in my case fixing this would either yield files that are several thousand lines of code long (which just breaks some IDEs like pycharm, which was the entire reason to have such a modular structure) or end up with some unimaginable mess just to support type checking (in which case the dev would probably *not* support the type checking to avoid breaking other design concerns, which leads to the code deterioration as a whole).

A compromise could be considered as having a config option that finds only import cycles that exist at runtime and can be stopped by being in the branches that always evaluate to false, so we have best of both worlds where you can find design issues separately and runtime issues separately 😊

**Comment by DetachHead** [comment]
thanks for the examples. these are tricky, and it's hard to suggest a full solution without more context for them, but maybe these circular type dependencies can be avoided if they were instead passed as generics?

example 1:
```py
# RootObject.py
class RootObject[T]:
    modules: list[T]
    # 500 lines of code here


# ModuleObject.py
class Module:  # this handles some specific functionality, not a generic class
    root: RootObject["Module"]
    # 500 lines of code here
```

example 2:
```py
# GenericObject.py
class GenericObject[Parent = "GenericObject", Child = "GenericObject"]:
    children: list[Child]
    parent: Parent

    def addChild(self, child: Child): ...



# ComplexState.py
class ComplexState(GenericObject['ComplexState', SubObject['ComplexState', GenericObject]]):
    def __init__(self):
        self.addChild(SubObject[ComplexState, GenericObject]())


# SubObject.py
class SubObject[Parent, Child](GenericObject[Parent, Child]):
    parent: Parent  # oh no!
```

doing so may increase type safety as well, though i realize doing stuff like this can make classes pretty complicated to work with, or it may not even work at all depending on what you're doing.

when i've encountered similar scenarios the solution that's worked for me has been to just define them all in one file ([the pyright codebase does this too](https://github.com/DetachHead/basedpyright/blob/main/packages/pyright-internal/src/parser/parseNodes.ts)). but as you mentioned, depending on how many classes you have / how large they are i realize this isn't always feasible.

regardless, i'm still reluctant to reduce the strictness of this rule because i've seen the consequences of people relying entirely on what's allowed at runtime to decide when circular dependencies are acceptable.

> A compromise could be considered as having a config option that finds only import cycles that exist at runtime and can be stopped by being in the branches that always evaluate to false, so we have best of both worlds where you can find design issues separately and runtime issues separately 😊

i guess we could add an `allowTypeOnlyImportCycles` option that's disabled by default and adequately documented warning about the dangers of enabling it. though i'd like to become more familiar with scenarios where they're unavoidable before adding such an option.

**Comment by Wizzerinus** [comment]
> when i've encountered similar scenarios the solution that's worked for me has been to just define them all in one file ([the pyright codebase does this too](https://github.com/DetachHead/basedpyright/blob/main/packages/pyright-internal/src/parser/parseNodes.ts)). but as you mentioned, depending on how many classes you have / how large they are i realize this isn't always feasible.

yeah this is a really good example, actually! this file is 3k lines because of the circular import problem, and such large files usually break IDEs (I had a 5k line file and after modifying it in Python you had to wait ~20 seconds for the Intellisense to fix it). So you kind of have to choose between "feasibility of file editing" and "have to find runtime errors manually/run all tests every time imports are edited/etc".

After thinking abt your examples I agree that it is probably useful to flag such issues in a lot of codebases, so having a flag like that is a better solution than modifying the importCycles thing by default

**Comment by DetachHead** [comment]
> yeah this is a really good example, actually! this file is 3k lines because of the circular import problem, and such large files usually break IDEs (I had a 5k line file and after modifying it in Pycharm you had to wait ~20 seconds for the Intellisense to update). So you kind of have to choose between "feasibility of file editing" and "have to find runtime errors manually/run all tests every time imports are edited/etc".

does the basedpyright language server have the same issue? i have some pretty large files in my projects too but it never seems to cause basedpyright to be noticably slower. if you aren't already, you may want to try the [pyright pycharm plugin](https://docs.basedpyright.com/latest/installation/ides/#pycharm) and see if that's any faster. though idk how to disable pycharm's existing python language features which may interfere with it.

**Comment by Wizzerinus** [comment]
I *think* basedpyright is fast in this regard but I don't know because I don't have huge files anymore :(

**Comment by DetachHead** [comment]
in that case i'm leaning towards closing this issue, but let us know if you do encounter any performance issues as a result of merging modules together to work around this issue and we can either re-open this or raise a separate issue

**Comment by Wizzerinus** [comment]
I'm sorry for bluntness, but this approach sounds kind of stupid honestly? there are people working on the same code base who use pycharm, so attempting to make basedpyright happy makes them unable to work on it :upside_down_face:

Not to mention this rule would tend devs towards monolithic code bases which sounds like the worst of both worlds D:

I am fine not enabling the rule for now, but yeah.

**Comment by DetachHead** [comment]
> I'm sorry for bluntness, but this approach sounds kind of stupid honestly? there are people working on the same code base who use pycharm, so attempting to make basedpyright happy makes them unable to work on it 🙃

i do try to keep in mind that people use basedpyright with all sorts of different editors, but pycharm uses its own outdated thing for language support instead of a language server, which in my experience is riddled with problems that make it basically unusable for any modern python development. for example [its type checking is infinitely worse than pyright and even mypy](https://youtrack.jetbrains.com/issue/PY-51358/Mypy-support-out-of-the-box).

i can re-open this issue and give it some more thought, but since circular dependencies are something i feel very strongly against i just don't feel right adding an option to change this behavior just because pycharm can't handle opening a file with a few thousand lines. its type checking is far less sophisticated than pyright so it really has no excuse to be that slow.

but maybe i'm being biased because i used to be a pycharm user. i paid for jetbrains IDEs for years and watched pycharm slowly rot since they haven't addressed any of these issues in years, barely kept up with new language features and still refuse to accept that language servers are the future (their LSP support is barebones and exclusive to the paid version)

> Not to mention this rule would tend devs towards monolithic code bases which sounds like the worst of both worlds D:

as i said earlier i believe when a codebase has separate modules that circularly depend on each other, it's still just one single monolithic module, but just split up into separate files to give the illusion that they are separate modules. this is the case with the codebase i deal with at work and it's an absolute nightmare to work with, it's just as difficult to untangle either way.

**Comment by indivisible** [comment]
I do see the point in warning on every circular import, but IMO there should be 2 separate rules: one that can trigger on any circular import (like it works now, but maybe as a warning), and one that only triggers an error on actually broken code. Or maybe mark the offending import line, and make it line-ignorable without disabling it altogether

fwiw you already need a TYPE_CHECKING guard on such imports so it's hard to miss that the code's "messy". forcing different hacks to satisfy a linter (like the generics idea) won't really make the design less tightly coupled, just obscures it a bit

**Comment by DetachHead** [comment]
> Or maybe mark the offending import line, and make it line-ignorable without disabling it altogether

yeah i wanna do that. see #827

> fwiw you already need a TYPE_CHECKING guard on such imports so it's hard to miss that the code's "messy".

not really sure what you mean. `TYPE_CHECKING` blocks don't always mean you're doing something messy. for example it could mean you just want to improve runtime performance by not importing modules that are only needed for type checking (see [these ruff rules](https://docs.astral.sh/ruff/rules/#flake8-type-checking-tc))

> forcing different hacks to satisfy a linter (like the generics idea) won't really make the design less tightly coupled, just obscures it a bit

i think using a `TYPE_CHECKING` import to avoid a runtime circular import error is even more hacky though. if two modules circularly depend on each other, they're essentially one big module anyway, and attempting to split them into separate files is just another way to obscure it.

the generics example i gave may not be a good solution here but it has been in some of the cases i've encountered. the correct solution to a circular dependency is a case by case basis and i was just suggesting ideas based on my experience without the full context for the problem.

**Comment by imLew** [comment]
I'm running into this issue myself right now while adding types to a codebase and I think that separating cases based on whether `TYPE_CHECKING`  is used or not would be the best near term solution.

1) Circular imports are fine for type checking, but runtime circular imports are obvious errors and should be caught. While it may be true that the first is a code smell, they are different issues. (With pydantic or other third party libraries circular type imports can become errors, but python will  always be dynamically typed.)
1) Adding type annotations to an existing project can be a lot of work. The current situation basically requires that users either turn off a useful error check or do a refactor.
1) Specific examples of this problem are a red herring as there are likely too many. Some might be solved relatively easily, but there is no general solution, so you run the risk of this becoming a whack-a-mole issue where people keep reporting slightly different cases.

However, to me, option 1 in #827 would also be a OK solution. Adding `if TYPE_CHECKING` is already a little clunky, so adding some comments to those lines wouldn't be too bad.

**Comment by DetachHead** [comment]
> 1. Circular imports are fine for type checking, but runtime circular imports are obvious errors and should be caught. While it may be true that the first is a code smell, they are different issues. 

this is a fair point, it probably is best to split them into separate rules.

> 2. Adding type annotations to an existing project can be a lot of work. The current situation basically requires that users either turn off a useful error check or do a refactor.

[baseline](https://docs.basedpyright.com/latest/benefits-over-pyright/baseline/) may be useful to you if you're updating an existing project

**Comment by oelhammouchi** [comment]
Hi, quick question: how should I go about disabling the warning for single case? I've tried putting a `# pyright: ignore[reportImportCycles]` comment everywhere but it doesn't seem to pick it up. I like to have the diagnostic in general, but in one particular case I'm dealing with having to import a version specifier from the `__init__.py` module to set it on an object while simultaneously exporting that object in the module (the import is deferred to a function, so no runtime issues).

(Thanks a lot for this amazing tool, I can't believe I stumbled across it so late, it's been a huge boon to my productivity)

**Comment by DetachHead** [comment]
hmm i don't think ignore comments work with `reportImportCycles` because the error is reported on the file itself rather than a specific line number. i guess the best solution is to just put a `# pyright: reportImportCycles=false` comment at the top of your file, which disables it for the whole file (which basically means the same thing since the error is always reported per-file).

IDEs seem to display the error as if it's being reported on line 1, so to prevent confusion i do think `# pyright:ignore` comments should work. i've opened #1312 to track this. also thank you for the kind words!

**Comment by vikigenius** [comment]
@DetachHead I do agree with the principle in general. But this often comes up when working with SqlAlchemy models and foreign keys.

app/db/models/chat_thread.py
```python
# -*- coding: utf-8 -*-
# pyright: reportImportCycles=false
from __future__ import annotations
from uuid import UUID
from typing import Any, TYPE_CHECKING
from advanced_alchemy.base import UUIDAuditBase
from sqlalchemy import ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship


class ChatThread(UUIDAuditBase):
    __tablename__ = 'chat_thread'
    __table_args__: dict[str, Any] = {"comment": "Chat Threads"}

    if TYPE_CHECKING:
        from app.db.models.user import User
        from app.db.models.chat_message import ChatMessage

    user_id: Mapped[UUID] = mapped_column(ForeignKey("user_account.id", ondelete="cascade"), nullable=False)
    title: Mapped[str | None] = mapped_column(nullable=True, default=None)

    user: Mapped['User'] = relationship(back_populates='chat_thread')
    messages: Mapped[list['ChatMessage']] = relationship(back_populates='thread')
```

app/db/models/chat_message.py

```python
# -*- coding: utf-8 -*-
# pyright: reportImportCycles=false
from __future__ import annotations
import enum
from uuid import UUID
from typing import Any, TYPE_CHECKING
from advanced_alchemy.base import UUIDAuditBase
from sqlalchemy import Enum, ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import JSONB


class Role(enum.Enum):
    """The message role."""
    user = 1
    assistant = 2


class ChatMessage(UUIDAuditBase):
    __tablename__ = 'chat_message'
    __table_args__: dict[str, Any] = {"comment": "Chat Messages"}

    if TYPE_CHECKING:
        from app.db.models.chat_thread import ChatThread

    chat_thread_id: Mapped[UUID] = mapped_column(ForeignKey("chat_thread.id", ondelete="cascade"), nullable=False)
    text: Mapped[str | None] = mapped_column(nullable=True, default=None)
    attributes: Mapped[dict[str, Any]] = mapped_column(JSONB, nullable=True, default=dict)
    role: Mapped[Role] = mapped_column(Enum(Role))
    thread: Mapped['ChatThread'] = relationship(back_populates='messages')
```

How would you get around this without disabling the cycles check?

**Comment by AldehilS** [comment]
@DetachHead I have the same problem that @vikigenius presented. It is pretty common in sqlalchemy to have your database models in different files, and establish the relationships between the two tables (this relationship is bidirectional) meaning that, if you want to use the typed notation of sqlalchemy, you must import fileA class into fileB and viceversa within the `if TYPE_CHECKING` block to prevent a circular import.
If you try to avoid importing the classes and only use strings for the type hints, you end up with tons of warnings from basedpyright because the type is not declared.

Maybe sqlalchemy models are one of the cases where `TYPE_CHECKING` is justified.

**Comment by DetachHead** [comment]
i'm not familiar with sqlalchemy, but i was already convinced by what @imLew said:

> Circular imports are fine for type checking, but runtime circular imports are obvious errors and should be caught. While it may be true that the first is a code smell, they are different issues.

for example, you may want to report a runtime circular import as an error, but a `TYPE_CHECKING` one as a warning. this would also allow you to disable the diagnostic on `TYPE_CHECKING` imports entirely if you want.

**Comment by tylerlaprade** [comment]
Hope I'm not beating a dead horse, but I run into this often with Django models' foreign key type definitions needing to be bidirectional, too.

**Comment by jprize** [comment]
> Hope I'm not beating a dead horse, but I run into this often with Django models' foreign key type definitions needing to be bidirectional, too.

Similar issue in Django with models defining managers (`objects = MyManager()`) but managers also needing to know the model types, easily fixed with `typing.TYPE_CHECKING`, but not working well with `basedpyright`.

**Comment by vhbui02** [comment]
Gosh I thought I was doing something wrong although I've put `from typing import TYPE_CHECKING`, `from __future__ import annotations` and wrap single quotes around the types, the error still persists. Maybe you need to update the diagnosis.

We could solve every circular dependencies via bubbling up the type. But, in my case, the types were too coupling that it's not worth to introduce an irregular. I'm working with Pydantic V2 also, it supports rebuild the model to keep runtime validation.

> [@DetachHead](https://github.com/DetachHead) quick question: how should I go about disabling the warning for single case? I've tried putting a `# pyright: ignore[reportImportCycles]` comment everywhere but it doesn't seem to pick it up. I like to have the diagnostic in general, but in one particular case I'm dealing with having to import a version specifier from the `__init__.py` module to set it on an object while simultaneously exporting that object in the module (the import is deferred to a function, so no runtime issues).

Exactly my case.

> hmm i don't think ignore comments work with `reportImportCycles` because the error is reported on the file itself rather than a specific line number. i guess the best solution is to just put a `# pyright: reportImportCycles=false` comment at the top of your file, which disables it for the whole file (which basically means the same thing since the error is always reported per-file).
> 
> IDEs seem to display the error as if it's being reported on line 1, so to prevent confusion i do think `# pyright:ignore` comments should work. i've opened [#1312](https://github.com/DetachHead/basedpyright/issues/1312) to track this. also thank you for the kind words!

Works perfectly. Thanks @DetachHead for creating such an awesome tool.

**Comment by thomas-mckay** [comment]
I was surprised by this too.

A quick MRE to illustrate:

in `registry.py`
```python
from __future__ import annotations
from .exceptions import AlreadyRegisteredError

class RegistryEntry:
    name: str

class Registry:
    _registry: set[RegistryEntry]
    def register(self, entry: RegistryEntry) -> None:
        if entry in self._registry:
            raise AlreadyRegisteredError(entry)
        self._registry.add(entry)
```

Then, in `exceptions.py`
```python
from __future__ import annotations   # Circular import detected.
from typing import TYPE_CHECKING
from my_lib.exceptions import MyLibError
if TYPE_CHECKING:
    from .registry import RegistryEntry

class RegistryError(MyLibError): ...

class AlreadyRegisteredError(RegistryError):
    def __init__(self, entry: RegistryEntry):
        super().__init__(f'{entry.name} is already registered.')
```

The error forces me to declare rich exceptions in the same file as the class that raises them. That's a bit of an arbitrary constraint for something that isn't violating any runtime rule. I happen to prefer declaring my exceptions in a per-package `exceptions.py` module and I prefer my exceptions to be rich (meaning they'll use types from the main module as inputs). This is not violating separation of concerns. I could even argue it's enforcing it.

One could argue putting `RegistryEntry` in its own module solves the circular dependency (it does), but it's not always practical (in my actual case, the common dependency is a 2-line type-alias that belongs with the registry definition). Certainly it's not something that should be imposed by the type-checker when the circular dependency is not a runtime error.
